### PR TITLE
Implementation for a configure widget

### PR DESCRIPTION
**Pull request by LittleWhite-tb**
I have worked on a new widget to configure Citra:

citra-configure-dialog

There is nothing in "Input" tab since we don't have the input widget yet.

citra-debug-configuration

Additionally, I have centralized all the configuration in Config class. Now, we should have one configure file only.

### CODE REVIEW

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-24T21:02:20Z: https://github.com/citra-emu/citra/compare/0b6cc0592dec5584d1a2211e9d777f0b2640faea...LittleWhite-tb:e6406a2c37402268cd0e799be9f7ed7a1bfe343e

File: src/citra_qt/configure_debug.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef CONFIGURE_DEBUG_H
+#define CONFIGURE_DEBUG_H
+
+#include <QWidget>
+
+namespace Ui {
+class ConfigureDebug;
+}
+
+class ConfigureDebug : public QWidget
```
Comment (lioncash) [inline src/citra_qt/configure_debug.h]: final class

File: src/citra_qt/configure_debug.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    Q_OBJECT
+
+public:
+    explicit ConfigureDebug(QWidget *parent = 0);
```
Comment (lioncash) [inline src/citra_qt/configure_debug.h]: nullptr

File: src/citra_qt/configure_dialog.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef CONFIGURE_DIALOG_H
+#define CONFIGURE_DIALOG_H
+
+#include <QDialog>
+
+namespace Ui {
+class ConfigureDialog;
+}
+
+class ConfigureDialog : public QDialog
```
Comment (lioncash) [inline src/citra_qt/configure_dialog.h]: final class

File: src/citra_qt/configure_dialog.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+{
+    Q_OBJECT
+
+public:
+    explicit ConfigureDialog(QWidget *parent = 0);
```
Comment (lioncash) [inline src/citra_qt/configure_dialog.h]: nullptr

File: src/citra_qt/configure_dialog.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    ~ConfigureDialog();
+
+    void applyConfiguration();
+
+private:
+    void setConfiguration();
+
+private:
+    Ui::ConfigureDialog *ui;
```
Comment (lioncash) [inline src/citra_qt/configure_dialog.h]: unique_ptr

File: src/citra_qt/configure_debug.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    ~ConfigureDebug();
+
+    void applyConfiguration();
+
+private:
+    void setConfiguration();
+
+private:
+    Ui::ConfigureDebug *ui;
```
Comment (lioncash) [inline src/citra_qt/configure_debug.h]: unique_ptr

File: src/citra_qt/configure_general.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef CONFIGURE_GENERAL_H
+#define CONFIGURE_GENERAL_H
+
+#include <QWidget>
+
+namespace Ui {
+class ConfigureGeneral;
+}
+
+class ConfigureGeneral : public QWidget
```
Comment (lioncash) [inline src/citra_qt/configure_general.h]: final class

File: src/citra_qt/configure_general.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+{
+    Q_OBJECT
+
+public:
+    explicit ConfigureGeneral(QWidget *parent = 0);
```
Comment (lioncash) [inline src/citra_qt/configure_general.h]: nullptr

File: src/citra_qt/configure_general.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    ~ConfigureGeneral();
+
+    void applyConfiguration();
+
+private:
+    void setConfiguration();
+
+private:
+    Ui::ConfigureGeneral *ui;
```
Comment (lioncash) [inline src/citra_qt/configure_general.h]: unique_ptr

File: src/citra_qt/configure_general.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#ifndef CONFIGURE_GENERAL_H
```
Comment (lioncash) [inline src/citra_qt/configure_general.h]: #pragma once

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-24T21:20:26Z: https://github.com/citra-emu/citra/compare/0b6cc0592dec5584d1a2211e9d777f0b2640faea...LittleWhite-tb:537cb9c8142873380d4b8c581afb69a2ff3c473a

File: src/citra_qt/configure_debug.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#ifndef CONFIGURE_DEBUG_H
```
Comment (lioncash) [inline src/citra_qt/configure_debug.h]: #pragma once

File: src/citra_qt/configure_dialog.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#ifndef CONFIGURE_DIALOG_H
```
Comment (lioncash) [inline src/citra_qt/configure_dialog.h]: #pragma once

File: src/citra_qt/ui_settings.h [hunk 13]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#ifndef UISETTINGS_H
```
Comment (lioncash) [inline src/citra_qt/ui_settings.h]: #pragma once

File: src/citra_qt/ui_settings.h [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <QByteArray>
+#include <QStringList>
+#include <QString>
+
+#include <vector>
+
+namespace UISettings {
+
+    typedef std::pair<QString, int> ContextedShortcut;
```
Comment (lioncash) [inline src/citra_qt/ui_settings.h]:
```cpp
using ContextedShortcut = std::pair<QString, int>;
using Shortcut = std::pair<QString, ContextedShortcut>;
```

Comment (lioncash) [inline src/citra_qt/ui_settings.h]: Also indentation should match the same level as the struct.

File: src/citra_qt/ui_settings.h [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+    QString roms_path;
+    QString symbols_path;
+    QString gamedir;
+    bool gamedir_deepscan;
+    QStringList recent_files;
+
+    // Shortcut name <Shortcut, context>
+    std::vector<Shortcut> shortcuts;
+} extern values;
```
Comment (lioncash) [inline src/citra_qt/ui_settings.h]: Please separate the variable declaration from the struct definition.

File: src/citra_qt/main.cpp [hunk 16]
```diff
@@ -494,26 +435,29 @@ void GMainWindow::ToggleWindowMode() {
 }
 
 void GMainWindow::OnConfigure() {
-    //GControllerConfigDialog* dialog = new GControllerConfigDialog(controller_ports, this);
+    ConfigureDialog configureDialog(this);
+    auto result = configureDialog.exec();
+    if ( result == QDialog::Accepted)
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: unnecessary space

**Comment by lioncash** [comment]
GDBStub UI contents shouldn't be smack in the center of the pane. If that's ever added to, then it'll look ugly.

File: src/citra_qt/config.cpp [hunk 17]
```diff
@@ -67,6 +67,50 @@ void Config::ReadValues() {
...
+
+    qt_config->beginGroup("Shortcuts");
+    QStringList groups = qt_config->childGroups();
+    for (auto group : groups) {
+        qt_config->beginGroup(group);
+
+        QStringList hotkeys = qt_config->childGroups();
+        for (auto hotkey : hotkeys) {
+            qt_config->beginGroup(hotkey);
+            UISettings::values.shortcuts.push_back(
```
Comment (lioncash) [inline src/citra_qt/config.cpp]: emplace_back

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-25T17:12:09Z: https://github.com/citra-emu/citra/compare/8b3994e9e4629423960f9ba22e5df01afd5b724b...LittleWhite-tb:555fc9ee1413d22949cada55a5e3512909d0ce01

**Comment by LittleWhite-tb** [comment]
All previous comments on code has been applied (final keyword, nullptr, pragma once...). Now, debug panel looks like this: 
configuration-debug

File: src/citra_qt/configure_debug.h [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    explicit ConfigureDebug(QWidget *parent = nullptr);
+    ~ConfigureDebug() {}
+
+    void applyConfiguration();
+
+private:
+    void setConfiguration();
+
+private:
+    Ui::ConfigureDebug ui;
```
Comment (lioncash) [inline src/citra_qt/configure_debug.h]: the unique_ptr method of instantiating these allows you to forward-declare the ui instance type.  Allowing you to remove the include from the header and move it to the .cpp file.

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-25T18:26:34Z: https://github.com/citra-emu/citra/compare/8b3994e9e4629423960f9ba22e5df01afd5b724b...LittleWhite-tb:cbe8e84964a6024acc8f2ee82e835220dd030ef5

**Comment by LittleWhite-tb** [comment]
Rework UI class to use std::unique_ptr.

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-14T19:25:51Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...LittleWhite-tb:3720310b110ec3cd032eeb7689c72658d7f1bb46

**Comment by LittleWhite-tb** [comment]
Finally, I have remove the final keyword for the three class I added. This is because the Travis Linux build was not successful : https://travis-ci.org/citra-emu/citra/jobs/104706605 (final keyword is not handled correctly by Qt 4 MOC).

**Comment by LittleWhite-tb** [comment]
Up, to solve issues with merging.

File: src/citra_qt/configure_general.ui [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+           <widget class="QCheckBox" name="toogle_hw_renderer">
+            <property name="text">
+             <string>Enable hardware renderer</string>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QCheckBox" name="toogle_shader_jit">
+            <property name="text">
+             <string>Enable Shader JIT</string>
```
Comment (JayFoxRox) [inline src/citra_qt/configure_general.ui]: "Enable shader JIT" ?

File: src/citra_qt/configure_debug.cpp [hunk 20]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "citra_qt/configure_debug.h"
+#include "ui_configure_debug.h"
+
+#include "core/core.h"
+#include "core/gdbstub/gdbstub.h" // TODO: can't include gdbstub without core.h
```
Comment (bunnei) [inline src/citra_qt/configure_debug.cpp]: Should this line be in `core.h`?

Comment (MerryMage) [inline src/citra_qt/configure_debug.cpp]: > `// TODO: can't include gdbstub without core.h`

Why?

If it's an issue with `PAddr` and `u32` not being defined in gdbstub.h, add `#include "common/common_types.h"` to gdbstub.h.

**Comment by LittleWhite-tb** [comment]
This branch will compile fine, when #1550 will be merged. Otherwise, the typo noticed by JayFoxRox is fixed and everything should be fine now.

**Comment by Kloen** [comment]
I'm wondering if all this new config related files should go under `citra-qt/config/` folder instead of just `citra-qt/`

**Comment by LittleWhite-tb** [comment]
From the actual code : 
> // TODO: Don't hardcode the path; let the frontend decide where to put the config files.
This can be an interesting PR to do :)

File: src/citra_qt/ui_settings.h [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+#include <QByteArray>
+#include <QStringList>
+#include <QString>
+
+#include <vector>
+
+namespace UISettings {
+
+using ContextedShortcut = std::pair<QString, int> ;
```
Comment (lioncash) [inline src/citra_qt/ui_settings.h]: Minor nit, but this should likely be `ContextualShortcut`

File: src/citra_qt/ui_settings.h [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+struct Values {
+    QByteArray geometry;
+    QByteArray state;
+
+    QByteArray renderwindow_geometry;
+
+    QByteArray gamelist_header_state;
+
+    QByteArray microprofile_geometry;
+    bool microprofile_visible;
```
Comment (lioncash) [inline src/citra_qt/ui_settings.h]: These bools should likely have default values.

Comment (LittleWhite-tb) [inline src/citra_qt/ui_settings.h]: The default value is set in Config::ReadValues() (which is call in the constructor).

**Comment by LittleWhite-tb** [comment]
Fix the name of the typedef, as noticed by @lioncash

**Comment by JayFoxRox** [comment]
As also noted on the merge commit: This causes giant slowdown because it seems to ignore the setting for HW / SW Rasterizer and always starts up as SW Rasterizer (even if the config window shows the Hardware renderer enabled).

I didn't check JIT/Interpreter yet.

**Comment by wwylele** [comment]
@JayFoxRox seems just config file broken. I tried enable and disable options several times and now they work fine

**Comment by JayFoxRox** [comment]
[This piece of code](https://github.com/citra-emu/citra/pull/1368/files#diff-5e8bf22f5c3b3579c402308dd96b7563L153) was just removed without adding anything which replaces it.

It's responsible to apply the JIT and HW settings on startup. If it's not done the config dialog will claim you are running the HW renderer when it really uses the SW rasterizer because VideoCore was never updated.

#1659 fixes it.

**Pull request merged**